lms/djangoapps/instructor/views/data_access.py

```diff
@@ -0,0 +1,374 @@
+    return {temp_query.id: students}  # pylint: disable=no-member
+
+
+def purge_temporary_queries():
+    """
+    Delete queries made more than 30 minutes ago along with the saved students from those queries
+    """
+    minutes_ago = datetime.datetime.now() - datetime.timedelta(minutes=TEMPORARY_QUERY_LIFETIME)
+    old_queries = TemporaryQuery.objects.filter(created__lt=minutes_ago)
+    saved_students = StudentsForQuery.objects.filter(query_id__in=old_queries.values_list(DatabaseFields.ID))
+    saved_students.delete()
+    old_queries.delete()
+
+
+@task(base=EmailWidgetTask)  # pylint: disable=not-callable
+def make_total_query(existing_queries):
+    """
+    Given individual queries that have already been made , aggregate students associated with those queries
+    """
+    aggregate_existing = set()
+    if len(existing_queries) != 0:
+        queryset = make_existing_query(existing_queries).values_list(
+            DatabaseFields.ID,
+            DatabaseFields.EMAIL,
+            DatabaseFields.PROFILE_NAME,
+        ).distinct()
+        for row in queryset:
+            aggregate_existing.add((row[0], row[1], row[2]))
+    return aggregate_existing
+
+
+def get_problem_users(course_id, query):
+    """
+    Gets Students filtered on specified problem-specific query criteria
+    """
+    results = []
+    if query.filter == ProblemFilters.OPENED:
+        results = open_query(course_id, query)
+    elif query.filter == ProblemFilters.COMPLETED:
+        results = completed_query(course_id, query)
+    elif query.filter == ProblemFilters.NOT_OPENED:
+        results = not_open_query(course_id, query)
+    elif query.filter == ProblemFilters.NOT_COMPLETED:
+        results = not_completed_query(course_id, query)
+    return results
+
+
+def get_section_users(course_id, query):
+    """
+    Gets Students filtered on specified section-specific query criteria
+    """
+    if query.filter == SectionFilters.OPENED:
+        results = open_query(course_id, query)
+    elif query.filter == SectionFilters.NOT_OPENED:
+        results = not_open_query(course_id, query)
+    return results
+
+
+def make_existing_query(existing_queries):
+    """
+    Aggregates single queries in a group into one unified set of students
+    """
+    query = User.objects
+    if existing_queries is None:
+        return None
+
+    query_dct = defaultdict(list)
+
+    for existing_query in existing_queries:
+        if existing_query == "" or existing_query == QueryStatus.WORKING:
+            continue
+        inclusion_type = TemporaryQuery.objects.filter(id=existing_query)
+        result = StudentsForQuery.objects.filter(query_id=existing_query).values_list(
+            DatabaseFields.STUDENT_ID,
+            flat=True,
+        )
+        if inclusion_type.exists():
+            query_dct[inclusion_type[0].inclusion].append(result)
+
+    for not_query in query_dct[INCLUSION_MAP.get(Inclusion.NOT)]:
+        query = query.exclude(id__in=not_query)
```

Comment:
for all these `id__in` cases, let's figure out if these become sql subqueries, or whether the subquery is eagerly executed and ids are passed in as a long list.